addons/portal/models/mail_thread.py

```diff
@@ -4,7 +4,52 @@
 import hashlib
 import hmac
 
-from odoo import fields, models, _
+from functools import wraps
+from inspect import Parameter, signature
+
+from odoo import fields, models, api, _
+from odoo.http import request
+from odoo.addons.bus.websocket import wsrequest
+from odoo.addons.portal.controllers.mail import _check_special_access
+
+
+def check_portal_access(func):
```

Comment:
You should probably avoid this kind of construct, as it makes code really hard to follow and understand, especially adding magic parameters. Why don't you improve the environment instead ? It seems to be the right place to store that kind of information, along user, lang, ... it seems more inlined with current code to avoid this at all cost.